### ISSUE TITLE
kubelet doesn't restart Pods on the hostNetwork if CNI isn't initialized

### ISSUE THREAD

**Issue by luxas** [kubernetes/kubernetes#43397]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): HEAD

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**: kubeadm:

How to reproduce on Ubuntu for example:

```
curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
cat <<EOF > /etc/apt/sources.list.d/kubernetes.list
deb http://apt.kubernetes.io/ kubernetes-xenial-unstable main
EOF
apt-get update && apt-get install -y docker.io kubeadm
kubeadm init --kubernetes-version v1.6.0-beta.4
```

**What happened**:

When something in a static Pod changes that requires the pod infra container to restart, like this command:
```
sed -e "s|spec:|spec:\n  dnsPolicy: ClusterFirstWithHostNet|" -i /etc/kubernetes/manifests/kube-apiserver.yaml
```

```
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.577643   26734 kuberuntime_gc.go:138] Failed to stop sandbox "640d0649f8ad383a2438cd17ee6f8b5b2a847462461e5e59987537778d604220" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.580730   26734 remote_runtime.go:109] StopPodSandbox "6c70b0c81f7c751b0abad687ca35a140b9c834c56d01a56b99020eaa2673e206" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-pdnfc_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.580745   26734 kuberuntime_gc.go:138] Failed to stop sandbox "6c70b0c81f7c751b0abad687ca35a140b9c834c56d01a56b99020eaa2673e206" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-pdnfc_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.596422   26734 remote_runtime.go:109] StopPodSandbox "a4a478447bea9939cc538db24737c47e50ebfcb80188c570e392e7c6e5b42d14" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.596443   26734 kuberuntime_gc.go:138] Failed to stop sandbox "a4a478447bea9939cc538db24737c47e50ebfcb80188c570e392e7c6e5b42d14" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.597515   26734 remote_runtime.go:109] StopPodSandbox "a8c81b8a3de45c5ee4dabcc7d2dcde6a85d5df896f51d83f483705140f576729" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-pwv8x_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.597535   26734 kuberuntime_gc.go:138] Failed to stop sandbox "a8c81b8a3de45c5ee4dabcc7d2dcde6a85d5df896f51d83f483705140f576729" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-pwv8x_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.624209   26734 remote_runtime.go:109] StopPodSandbox "df9d016b47220b5d56bce5d71f8e607be0f0fee73d95a83f14247b0fddc1221f" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-vqvrq_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.624239   26734 kuberuntime_gc.go:138] Failed to stop sandbox "df9d016b47220b5d56bce5d71f8e607be0f0fee73d95a83f14247b0fddc1221f" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-vqvrq_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.626239   26734 remote_runtime.go:109] StopPodSandbox "e6be13d131e76024bb6633f113f44215e9ed375af332233b0fcdc02ec8a0da38" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-k3vj8_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.626271   26734 kuberuntime_gc.go:138] Failed to stop sandbox "e6be13d131e76024bb6633f113f44215e9ed375af332233b0fcdc02ec8a0da38" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-k3vj8_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.627350   26734 remote_runtime.go:109] StopPodSandbox "edc38912ecb02b7d06607d95f45a7158123a62c61c32ae1de2135c5c45646f22" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-hq9gg_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.627367   26734 kuberuntime_gc.go:138] Failed to stop sandbox "edc38912ecb02b7d06607d95f45a7158123a62c61c32ae1de2135c5c45646f22" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "self-hosted-kube-apiserver-hq9gg_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.631330   26734 remote_runtime.go:109] StopPodSandbox "f6a61129c8231043080728d7c13dcb32860c52d9b0ac2926f549614120170934" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 20 17:40:49 rook-test kubelet-patched[26734]: E0320 17:40:49.631341   26734 kuberuntime_gc.go:138] Failed to stop sandbox "f6a61129c8231043080728d7c13dcb32860c52d9b0ac2926f549614120170934" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
```

**What you expected to happen**:

I expected the apiserver Pod to restart as usual (it works when disabling CRI)

**How to reproduce it** (as minimally and precisely as possible):

Described above

I think the change here will be pretty minimal, see: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_sandbox.go#L119

The problem is probably here somewhere: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_sandbox.go#L181

This is pretty serious, I would maybe consider it blocking for the v1.6 release.
cc @kubernetes/release-team 

@kubernetes/sig-network-misc @kubernetes/sig-node-bugs @yujuhong

**Comment by thockin** [comment]
#43368 ?

@yujuhong PTAL, I agree this could be a blocker

@bowei

**Comment by yujuhong** [comment]
> @yujuhong PTAL, I agree this could be a blocker

Already looking. Trying to reproduce the issue.

dockershim should not be calling teardown for a pod with host network. I suspect `PodSandboxStatus` returning an error (for whatever reason), but need to confirm.

**Comment by bowei** [comment]
Taking a look at the kubeadm repro

**Comment by yujuhong** [comment]
I ran into other issues while trying to reproduce this, but I never got the "cni config uninitialized" to show up...

The main issue @bowei and I ran into was that kubelet did not restart the static pod after we modified the manifest file. This was likely caused by inotify events not being processed properly (or been dropped). I've asked @Random-Liu to help look into the issue. 

For static pods, any change to the pod spec should result in a completely new pod UID, i.e., it should be treated as a new pod. The creation of the new pod should not be blocked by the teardown of the old pod. This is strange.

**Comment by yujuhong** [comment]
> The main issue @bowei and I ran into was that kubelet did not restart the static pod after we modified the manifest file. This was likely caused by inotify events not being processed properly (or been dropped). I've asked @Random-Liu to help look into the issue.

This seems to have been caused mostly by modifying the manifest file using an editor (which generates temporary files). I am going to bypass this issue first by copying and replacing the file.

> For static pods, any change to the pod spec should result in a completely new pod UID, i.e., it should be treated as a new pod. The creation of the new pod should not be blocked by the teardown of the old pod. This is strange.

Apparently kubelet treats some changes as UPDATE operations, instead of truly a new pod. Looking into this as well.

**Comment by yujuhong** [comment]
> Apparently kubelet treats some changes as UPDATE operations, instead of truly a new pod. Looking into this as well.

...... @Random-Liu traced the code path and found that we use the name/namespace to identify the pod in an internal cache. This causes some confusion when editing the pod. We'll fix that first.
/cc @dchen1107

**Comment by yujuhong** [comment]
@luxas @bowei could you see if you can reproduce the issue with this kubelet patch #43413?

I've tried and not once have I encountered the cni configuration issue as reported in the original post.

**Comment by yujuhong** [comment]
Both @bowei and I cannot reproduce this issue. @luxas could you try again and let me know if it's reproducible for you? Thanks!

BTW, don't edit the manifest file directly. There are some known issues. If you copy, modify, and replace the manifest file, it should be ok

**Comment by luxas** [comment]
What I ran now:
```console
$ kubeadm init --kubernetes-version v1.6.0-beta.4
$ sed -e "s|spec:|spec:\n  dnsPolicy: ClusterFirstWithHostNet|" -i /etc/kubernetes/manifests/kube-apiserver.yaml
$ docker ps
CONTAINER ID        IMAGE                                                                                                                            COMMAND                  CREATED              STATUS              PORTS               NAMES
dd2b270f6da7        gcr.io/google_containers/kube-apiserver-amd64@sha256:40fece6b61de6e30bc052238d10b000a25acac1b86e7cd3452551f00a7290149            "kube-apiserver --sto"   About a minute ago   Up About a minute                       k8s_kube-apiserver_kube-apiserver-rook-test_kube-system_63023b4160a81cee15d89a745ccfa390_0
c284d3aefb2f        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 About a minute ago   Up About a minute                       k8s_POD_kube-apiserver-rook-test_kube-system_63023b4160a81cee15d89a745ccfa390_0
7f7de095b3c5        gcr.io/google_containers/kube-proxy-amd64@sha256:b2c90eb638b94b9441d7d19d0076988dcbcc966be27d539c1ad8cd8d9f16466d                "/usr/local/bin/kube-"   3 minutes ago        Up 3 minutes                            k8s_kube-proxy_kube-proxy-lgn7s_kube-system_25672393-0e46-11e7-88da-42010a840002_0
18dd37631cb4        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 3 minutes ago        Up 3 minutes                            k8s_POD_kube-proxy-lgn7s_kube-system_25672393-0e46-11e7-88da-42010a840002_0
6c4934769ab2        gcr.io/google_containers/kube-scheduler-amd64@sha256:d942e64fe6a197ea7c68c969ab94b39809a4b0d3de78b7f0437c8fc6f69df45a            "kube-scheduler --lea"   3 minutes ago        Up 3 minutes                            k8s_kube-scheduler_kube-scheduler-rook-test_kube-system_a1b97953b2faec088924b52f69d818b4_0
7fe4803e945b        gcr.io/google_containers/kube-controller-manager-amd64@sha256:9f90a5d01c8e2e5e00e9c345244cb35be67e069fabfe11838ac089eae41711c6   "kube-controller-mana"   3 minutes ago        Up 3 minutes                            k8s_kube-controller-manager_kube-controller-manager-rook-test_kube-system_813b655994cf75404fde4088577eef5a_0
e51354759bb8        gcr.io/google_containers/etcd-amd64@sha256:d83d3545e06fb035db8512e33bd44afb55dea007a3abd7b17742d3ac6d235940                      "etcd --listen-client"   3 minutes ago        Up 3 minutes                            k8s_etcd_etcd-rook-test_kube-system_7075157cfd4524dbe0951e00a8e3129e_0
e4a7b6bc36b6        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 3 minutes ago        Up 3 minutes                            k8s_POD_kube-scheduler-rook-test_kube-system_a1b97953b2faec088924b52f69d818b4_0
6c3d5bd58ee9        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 3 minutes ago        Up 3 minutes                            k8s_POD_kube-controller-manager-rook-test_kube-system_813b655994cf75404fde4088577eef5a_0
51926ae4c609        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 3 minutes ago        Up 3 minutes                            k8s_POD_etcd-rook-test_kube-system_7075157cfd4524dbe0951e00a8e3129e_0
```

So it seems like this issue is fixed, :+1: the apiserver did now restart correctly. Earlier it didn't get restarted at all; it just remained in its initial state.

Still there are some somewhat confusing logs, I don't think the CNI plugin should ever be called when dealing with Pods with HostNetwork, but it seems like kubelet calls CNI regardless of networking type right now...

I think that shold be fixed in the future, but that's not a blocker since it seems to work now
Can you create an issue for not calling the network plugin for hostNetwork Pods so we don't lose it?
I guess you have a little bit more context you can add there...

```
Mar 21 14:55:41 rook-test kubelet-patched[18578]: W0321 14:55:41.337881   18578 kubelet.go:1524] Deleting mirror pod "kube-apiserver-rook-test_kube-system(205ca46a-0e46-11e7-88da-42010a840002)" because it is outdated
Mar 21 14:55:41 rook-test kubelet-patched[18578]: W0321 14:55:41.338825   18578 kubelet.go:1524] Deleting mirror pod "kube-apiserver-rook-test_kube-system(205ca46a-0e46-11e7-88da-42010a840002)" because it is outdated
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.358296   18578 mirror_client.go:88] Failed deleting a mirror pod "kube-apiserver-rook-test_kube-system": read tcp 10.132.0.2:57614->10.132.0.2:6443: read: connection reset by peer
Mar 21 14:55:41 rook-test kubelet-patched[18578]: W0321 14:55:41.358929   18578 status_manager.go:465] Failed to update status for pod "_()": Get https://10.132.0.2:6443/api/v1/namespaces/kube-system/pods/kube-apiserver-rook-test: read tcp 10.132.0.2:57614->10.132.0.2:6443: read: connection reset by peer
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.358980   18578 mirror_client.go:88] Failed deleting a mirror pod "kube-apiserver-rook-test_kube-system": read tcp 10.132.0.2:57614->10.132.0.2:6443: read: connection reset by peer
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.359209   18578 reflector.go:304] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to watch *v1.Pod: Get https://10.132.0.2:6443/api/v1/pods?fieldSelector=spec.nodeName%3Drook-test&resourceVersion=312&timeoutSeconds=521&watch=true: dial tcp 10.132.0.2:6443: getsockopt: connection refused
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.359238   18578 kubelet.go:1535] Failed creating a mirror pod for "kube-apiserver-rook-test_kube-system(63023b4160a81cee15d89a745ccfa390)": Post https://10.132.0.2:6443/api/v1/namespaces/kube-system/pods: dial tcp 10.132.0.2:6443: getsockopt: connection refused
Mar 21 14:55:41 rook-test kubelet-patched[18578]: W0321 14:55:41.359545   18578 status_manager.go:465] Failed to update status for pod "_()": Get https://10.132.0.2:6443/api/v1/namespaces/kube-system/pods/kube-apiserver-rook-test: dial tcp 10.132.0.2:6443: getsockopt: connection refused
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.359626   18578 kubelet.go:1535] Failed creating a mirror pod for "kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)": Post https://10.132.0.2:6443/api/v1/namespaces/kube-system/pods: dial tcp 10.132.0.2:6443: getsockopt: connection refused
Mar 21 14:55:41 rook-test kubelet-patched[18578]: I0321 14:55:41.359793   18578 kuberuntime_manager.go:458] Container {Name:kube-apiserver Image:gcr.io/google_containers/kube-apiserver-amd64:v1.6.0-beta.4 Command:[kube-apiserver --storage-backend=etcd3 --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --service-account-key-file=/etc/kubernetes/pki/sa.pub --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --insecure-port=0 --experimental-bootstrap-token-auth=true --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --secure-port=6443 --requestheader-username-headers=X-Remote-User --client-ca-file=/etc/kubernetes/pki/ca.crt --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds --allow-privileged=true --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --requestheader-group-headers=X-Remote-Group --service-cluster-ip-range=10.96.0.0/12 --tls-private-key-file=/etc/kubernetes/pki/apiserver.key --authorization-mode=RBAC --advertise-address=10.132.0.2 --etcd-servers=http://127.0.0.1:2379] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/healthz,Port:6443,Host:127.0.0.1,Scheme:HTTPS,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:15,TimeoutSeconds:15,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:8,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Mar 21 14:55:41 rook-test kubelet-patched[18578]: I0321 14:55:41.659624   18578 kuberuntime_manager.go:458] Container {Name:kube-apiserver Image:gcr.io/google_containers/kube-apiserver-amd64:v1.6.0-beta.4 Command:[kube-apiserver --storage-backend=etcd3 --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --service-account-key-file=/etc/kubernetes/pki/sa.pub --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --insecure-port=0 --experimental-bootstrap-token-auth=true --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --secure-port=6443 --requestheader-username-headers=X-Remote-User --client-ca-file=/etc/kubernetes/pki/ca.crt --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota,DefaultTolerationSeconds --allow-privileged=true --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --requestheader-group-headers=X-Remote-Group --service-cluster-ip-range=10.96.0.0/12 --tls-private-key-file=/etc/kubernetes/pki/apiserver.key --authorization-mode=RBAC --advertise-address=10.132.0.2 --etcd-servers=http://127.0.0.1:2379] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[cpu:{i:{value:250 scale:-3} d:{Dec:<nil>} s:250m Format:DecimalSI}]} VolumeMounts:[{Name:k8s ReadOnly:true MountPath:/etc/kubernetes/ SubPath:} {Name:certs ReadOnly:false MountPath:/etc/ssl/certs SubPath:}] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/healthz,Port:6443,Host:127.0.0.1,Scheme:HTTPS,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:15,TimeoutSeconds:15,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:8,} ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we s
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.732809   18578 remote_runtime.go:86] RunPodSandbox from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.732883   18578 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)" failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.732900   18578 kuberuntime_manager.go:619] createPodSandbox for pod "kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)" failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 21 14:55:41 rook-test kubelet-patched[18578]: E0321 14:55:41.732958   18578 pod_workers.go:182] Error syncing pod b6f5ab4a84ad8cfa02fde7a4aeb1ea45 ("kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)"), skipping: failed to "CreatePodSandbox" for "kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)" with CreatePodSandboxError: "CreatePodSandbox for pod \"kube-apiserver-rook-test_kube-system(b6f5ab4a84ad8cfa02fde7a4aeb1ea45)\" failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod \"kube-apiserver-rook-test_kube-system\" network: cni config uninitialized"
Mar 21 14:55:41 rook-test kubelet-patched[18578]: W0321 14:55:41.893179   18578 docker_sandbox.go:263] NetworkPlugin cni failed on the status hook for pod "kube-apiserver-rook-test_kube-system": Unexpected command output nsenter: cannot open : No such file or directory
Mar 21 14:55:42 rook-test kubelet-patched[18578]: E0321 14:55:42.362444   18578 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to list *v1.Pod: Get https://10.132.0.2:6443/api/v1/pods?fieldSelector=spec.nodeName%3Drook-test&resourceVersion=0: dial tcp 10.132.0.2:6443: getsockopt: connection refused
Mar 21 14:55:42 rook-test kubelet-patched[18578]: W0321 14:55:42.908393   18578 kubelet.go:1524] Deleting mirror pod "kube-apiserver-rook-test_kube-system(205ca46a-0e46-11e7-88da-42010a840002)" because it is outdated
Mar 21 14:55:42 rook-test kubelet-patched[18578]: E0321 14:55:42.922857   18578 mirror_client.go:88] Failed deleting a mirror pod "kube-apiserver-rook-test_kube-system": User "system:node:rook-test" cannot delete pods in the namespace "kube-system". (delete pods kube-apiserver-rook-test)
Mar 21 14:55:42 rook-test kubelet-patched[18578]: W0321 14:55:42.922862   18578 status_manager.go:465] Failed to update status for pod "_()": User "system:node:rook-test" cannot get pods in the namespace "kube-system". (get pods kube-apiserver-rook-test)
Mar 21 14:55:42 rook-test kubelet-patched[18578]: E0321 14:55:42.923670   18578 kubelet.go:1535] Failed creating a mirror pod for "kube-apiserver-rook-test_kube-system(63023b4160a81cee15d89a745ccfa390)": User "system:node:rook-test" cannot create pods in the namespace "kube-system". (post pods)
Mar 21 14:55:43 rook-test kubelet-patched[18578]: W0321 14:55:43.045441   18578 status_manager.go:465] Failed to update status for pod "_()": User "system:node:rook-test" cannot get pods in the namespace "kube-system". (get pods kube-apiserver-rook-test)
Mar 21 14:55:43 rook-test kubelet-patched[18578]: W0321 14:55:43.046297   18578 status_manager.go:465] Failed to update status for pod "_()": User "system:node:rook-test" cannot get pods in the namespace "kube-system". (get pods kube-apiserver-rook-test)
Mar 21 14:55:43 rook-test kubelet-patched[18578]: W0321 14:55:43.047072   18578 status_manager.go:465] Failed to update status for pod "_()": User "system:node:rook-test" cannot get pods in the namespace "kube-system". (get pods kube-apiserver-rook-test)
Mar 21 14:55:43 rook-test kubelet-patched[18578]: E0321 14:55:43.363262   18578 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to list *v1.Pod: User "system:node:rook-test" cannot list pods at the cluster scope. (get pods)
Mar 21 14:55:43 rook-test kubelet-patched[18578]: W0321 14:55:43.913040   18578 kubelet.go:1524] Deleting mirror pod "kube-apiserver-rook-test_kube-system(205ca46a-0e46-11e7-88da-42010a840002)" because it is outdated
Mar 21 14:55:44 rook-test kubelet-patched[18578]: W0321 14:55:44.916977   18578 kubelet.go:1524] Deleting mirror pod "kube-apiserver-rook-test_kube-system(205ca46a-0e46-11e7-88da-42010a840002)" because it is outdated
Mar 21 14:56:13 rook-test kubelet-patched[18578]: W0321 14:56:13.219370   18578 docker_sandbox.go:263] NetworkPlugin cni failed on the status hook for pod "kube-apiserver-rook-test_kube-system": Unexpected command output nsenter: cannot open : No such file or directory
Mar 21 14:56:13 rook-test kubelet-patched[18578]: E0321 14:56:13.220090   18578 remote_runtime.go:109] StopPodSandbox "040f1aa57fe7a059ff04dcee80c95d0d4888d8bf677899472e9fb3207647bd60" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
Mar 21 14:56:13 rook-test kubelet-patched[18578]: E0321 14:56:13.220120   18578 kuberuntime_gc.go:138] Failed to stop sandbox "040f1aa57fe7a059ff04dcee80c95d0d4888d8bf677899472e9fb3207647bd60" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-apiserver-rook-test_kube-system" network: cni config uninitialized
```

**Comment by yujuhong** [comment]
@luxas I actually never seen the "cni" error in all my attempts to reproduce this issue yesterday. Could you make sure you run this on a new machine (with no stale state in `/var/lib/`) and with the latest kubelet from HEAD? If you still see this error, feel free to open a new issue and we'll look into it more.

**Comment by kensimon** [comment]
I'm seeing this on v1.6.0 using kubeadm on a new cluster with the latest CNI release, with Docker 1.12.6.

The workflow I'm going through for installation is basically:

- Install kubelet/kubeadm/cni/docker/etc
- `kubeadm init` with the following yaml:
```
---
apiVersion: kubeadm.k8s.io/v1alpha1
kind: MasterConfiguration
token: <token_here>
cloudProvider: aws
kubernetesVersion: v1.6.0
apiServerCertSANs:
- <lb_here>.us-west-2.elb.amazonaws.com
```

And I watch the kubelet logs and see:

```
Mar 29 00:05:48 ip-10-0-30-185 kubelet[20078]: E0329 00:05:48.272213   20078 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Mar 29 00:05:53 ip-10-0-30-185 kubelet[20078]: W0329 00:05:53.273076   20078 cni.go:157] Unable to update cni config: No networks found in /etc/cni/net.d
... repeat a lot ...
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: W0329 00:06:33.709686   20078 docker_sandbox.go:263] NetworkPlugin cni failed on the status hook for pod "kube-dns-3913472980-j6nx1_kube-system": Unexpected command output nsenter: cannot open : No such file or directory
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]:  with error: exit status 1
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: E0329 00:06:33.710478   20078 remote_runtime.go:109] StopPodSandbox "b3f495241229d75c3477e6f0b72da3311f207c740079810feac4a023c8432451" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-dns-3913472980-j6nx1_kube-system" network: cni config uninitialized
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: E0329 00:06:33.710496   20078 kuberuntime_gc.go:138] Failed to stop sandbox "b3f495241229d75c3477e6f0b72da3311f207c740079810feac4a023c8432451" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-dns-3913472980-j6nx1_kube-system" network: cni config uninitialized
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: W0329 00:06:33.715109   20078 docker_sandbox.go:263] NetworkPlugin cni failed on the status hook for pod "kube-dns-3913472980-9rm20_kube-system": Unexpected command output nsenter: cannot open : No such file or directory
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]:  with error: exit status 1
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: E0329 00:06:33.715963   20078 remote_runtime.go:109] StopPodSandbox "4460f10614d96437692f4409a0931059ced38a6bd3804732ca6dea4f66cca22d" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-dns-3913472980-9rm20_kube-system" network: cni config uninitialized
Mar 29 00:06:33 ip-10-0-30-185 kubelet[20078]: E0329 00:06:33.715984   20078 kuberuntime_gc.go:138] Failed to stop sandbox "4460f10614d96437692f4409a0931059ced38a6bd3804732ca6dea4f66cca22d" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-dns-3913472980-9rm20_kube-system" network: cni config uninitialized
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.090181   20078 kuberuntime_manager.go:844] PodSandboxStatus of sandbox "39e3d1c1ad9fd04387dd2f8c2ff3ea3eef9a2d93899d3056a746e3d75bf80b5e" for pod "kube-scheduler-ip-10-0-30-185.us-west-2.compute.internal_kube-system(7f471661c3dcdd37ce5d62477f224ab9)" error: rpc error: code = 2 desc = Error response from daemon: {"message":"devmapper: Unknown device 0a0ae56d9308ec76a3c3533aecf2664dafd9875275db03885fb72829f018013b"}
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.090221   20078 generic.go:239] PLEG: Ignoring events for pod kube-scheduler-ip-10-0-30-185.us-west-2.compute.internal/kube-system: rpc error: code = 2 desc = Error response from daemon: {"message":"devmapper: Unknown device 0a0ae56d9308ec76a3c3533aecf2664dafd9875275db03885fb72829f018013b"}
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091142   20078 docker_sandbox.go:205] Failed to stop sandbox "00409d5c02d9192636e042b6accc8473ec9a387a844d3872fb7d98245df4d635": Error response from daemon: {"message":"No such container: 00409d5c02d9192636e042b6accc8473ec9a387a844d3872fb7d98245df4d635"}
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091202   20078 remote_runtime.go:109] StopPodSandbox "00409d5c02d9192636e042b6accc8473ec9a387a844d3872fb7d98245df4d635" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-controller-manager-ip-10-0-30-185.us-west-2.compute.internal_kube-system" network: cni config uninitialized
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091215   20078 kuberuntime_gc.go:138] Failed to stop sandbox "00409d5c02d9192636e042b6accc8473ec9a387a844d3872fb7d98245df4d635" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-controller-manager-ip-10-0-30-185.us-west-2.compute.internal_kube-system" network: cni config uninitialized
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091872   20078 docker_sandbox.go:205] Failed to stop sandbox "0c91e6c33ed5ce20f427643e62ced54f0011fca0ade77e4b013d62f868ebb85a": Error response from daemon: {"message":"No such container: 0c91e6c33ed5ce20f427643e62ced54f0011fca0ade77e4b013d62f868ebb85a"}
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091926   20078 remote_runtime.go:109] StopPodSandbox "0c91e6c33ed5ce20f427643e62ced54f0011fca0ade77e4b013d62f868ebb85a" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-scheduler-ip-10-0-30-185.us-west-2.compute.internal_kube-system" network: cni config uninitialized
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.091939   20078 kuberuntime_gc.go:138] Failed to stop sandbox "0c91e6c33ed5ce20f427643e62ced54f0011fca0ade77e4b013d62f868ebb85a" before removing: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-scheduler-ip-10-0-30-185.us-west-2.compute.internal_kube-system" network: cni config uninitialized
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.092742   20078 docker_sandbox.go:205] Failed to stop sandbox "144f4bceaca3613aef9aef2323e891032895a7ad4a1493445066dcc73c0c5934": Error response from daemon: {"message":"No such container: 144f4bceaca3613aef9aef2323e891032895a7ad4a1493445066dcc73c0c5934"}
Mar 29 00:06:34 ip-10-0-30-185 kubelet[20078]: E0329 00:06:34.092798   20078 remote_runtime.go:109] StopPodSandbox "144f4bceaca3613aef9aef2323e891032895a7ad4a1493445066dcc73c0c5934" from runtime service failed: rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod "kube-scheduler-ip-10-0-30-185.us-west-2.compute.internal_kube-system" network: cni config uninitialized
... repeats a lot ...
```

Ultimately kubelet logs these two lines in succession repeatedly:

```
Mar 29 00:17:58 ip-10-0-30-185 kubelet[20078]: W0329 00:17:58.433924   20078 cni.go:157] Unable to update cni config: No networks found in /etc/cni/net.d
Mar 29 00:17:58 ip-10-0-30-185 kubelet[20078]: E0329 00:17:58.434248   20078 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```

while the node stays in a state of `NotReady` perpetually.  So I think this is the same bug.

kubeadm's manifests are all defining files with `hostNetwork: true`, and I don't have any other pods (this is a fresh installation), which suggests to me that the initial components laid down by kubeadm are supposed to work without CNI set up yet but they're not for some reason.

If I install calico networking with `kubectl apply -f http://docs.projectcalico.org/v2.1/getting-started/kubernetes/installation/hosted/kubeadm/calico.yaml`, things move past this, but there's other calico bugs on 1.6 preventing me from getting much farther.

Seems like pods requiring CNI setup even if they're `hostNetwork` is preventing the initialization of a new cluster, at least for me.

**Comment by kensimon** [comment]
^^ cc @luxas @jbeda @thockin @yujuhong

**Comment by thockin** [comment]
@bowei @freehan

**Comment by yujuhong** [comment]
> Seems like pods requiring CNI setup even if they're hostNetwork is preventing the initialization of a new cluster, at least for me.

I just tested this with kubeadm+v1.6.0, and my experience was a bit different from what you described.

1. `kubeadm init` properly started kubelet and all master components.
2. All master components are running, but node/kubelet stays "NotReady".
3. `kubeadm` continues waiting for the node to become ready, and would not move on to the next phase to create kube-proxy/-dns addons.

For (2), the node would stay "not ready" until the network is properly configured to prevent the scheduler from assigning any pod to the node. I believe this is working as intended (introduced in #43474).

As for (3), if cni configuration would only be populated after the node becomes ready (assuming this is the case, correct me if I am wrong @luxas @jbeda), the setup process would be inherently flawed because of the circular dependency of (2) and (3). 

@kensimon hostnetwork pods should run regardless of node being ready or not. If you don't see all the master components running and ready (`kubectl get pods --all-namespaces`), could you run `kubelet version` to verify kubelet's version?

**Comment by kensimon** [comment]
@yujuhong thanks for looking into this.  Sounds like we're seeing the same thing, I may have been misattributing the `NotReady` state to pods not coming up when that's not the case.  Sounds like it may be a kubeadm bug then?

```
# kubeadm version
kubeadm version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:24:30Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
# kubelet --version
Kubernetes v1.6.0
```

I can get this to reproduce reliably using the setup script here: https://github.com/kensimon/aws-quickstart/blob/3a811d00982d9c265e2f00f93d07731f89784b95/scripts/setup-k8s-master.sh.in, although it's nothing really more complicated than just running `kubeadm init` on a new cluster.

**Comment by kensimon** [comment]
^^ looks like #43815 is the real bug here, I'll move my conversation there.

**Comment by jmillikin-stripe** [comment]
Can this be reopened? Regardless of whether kubeadm is doing the wrong thing, kubelet should not fail to delete a `hostNetworking` pod just because the CNI state is broken.
